docs/contributing/999-rfcs/2024-01-29-tests.md

```diff
@@ -0,0 +1,298 @@
+---
+title: RFC Template
+description: Template for creating a new RFC
+---
+
+# Tests
+
+- **Author(s):**: @tsuf239
+- **Submission Date**: {YYYY-MM-DD}
+- **Stage**: draft
+- **Stage Date**: 2024-01-30
+
+This rfc lists a few features that improve test-writing in wing, and strengthen wing's test framework.
+
+1. [scopes](#1-scopes)
+2. [lifecycle hooks](#2-lifecycle-hooks)
+3. [integration with external libs](#3-integration-with-external-libs)
+4. [mocks](#4-mocks)
+5. [spies](#5-spies)
+6. [resource management optimization](#6-resource-management-optimization)
+
+---
+
+## 1. Scopes
+
+### Design
+
+Currently there is a single test scope:
+
+```
+test {
+
+}
+```
+
+Each test run in separation, and the environment (the collection of resources available in the file) is duplicated per test. Since this is the only way available now, the user cannot re-use resources between tests. The test are running on parallel- one on each cloud.Function and each reports separately to the test runner.
+
+#### Suggested:
+
+We'll add another scope (and have two in total- outer scope, and inner scope) for the following structure:
+//TODO: names
+
+```js
+// file/global scope - preflight code only. Resources initialized here will be duplicated across outerScopes.
+let counter = new cloud.Counter();
+
+outerScope "bucket tests" {
+  /// preflight code only. Resources initialized here will be shared across innerScopes.
+  let bucket = new cloud.Bucket();
+
+  innerScope "bucket starts empty" {
```

Comment:
When this topic has been brought up, usually there's mention of a preflight/inflight test concept, where the preflight test is the setup for the inflight tests. WDYT?

```wing
// breaking change, but like other declarations the phase can be inferred (preflight)
test "bucket tests" {
  /// preflight code only. Resources initialized here will be shared across tests here
  let bucket = new cloud.Bucket();

  // explicitly inflight tests within the preflight test
  inflight test "bucket starts empty" {
  }
}
```

It's possible that users will be confused that it's a test within a test, although with good test output maybe it would be very intuitive. 
If it does seem confusing, maybe we can be more explicit about it being a setup or group. Just some ideas:

```wing
suite "bucket tests" {
  /// preflight code only. Resources initialized here will be shared across tests here
  let bucket = new cloud.Bucket();
  
  preflight test "bucket as prop" {
  }

  // explicitly inflight tests within the preflight test
  inflight test "bucket starts empty" {
  }
}
```

```wing
test app "bucket tests" {
  /// preflight code only. Resources initialized here will be shared across tests here
  let bucket = new cloud.Bucket();
  
  preflight test "bucket as prop" {
  }

  // explicitly inflight tests within the preflight test
  inflight test "bucket starts empty" {
  }
}
```